cargo-near/src/abi.rs

```diff
@@ -32,19 +33,32 @@ pub(crate) fn generate_abi(
     let original_cargo_toml: toml::value::Table =
         toml::from_slice(&fs::read(&crate_metadata.manifest_path.path)?)?;
 
-    if !original_cargo_toml
+    let near_sdk_dep = original_cargo_toml
         .get("dependencies")
-        .ok_or_else(|| anyhow::anyhow!("[dependencies] section not found"))?
-        .get("near-sdk")
-        .ok_or_else(|| anyhow::anyhow!("near-sdk dependency not found"))?
-        .as_table()
-        .ok_or_else(|| anyhow::anyhow!("near-sdk dependency should be a table"))?
-        .get("features")
-        .and_then(|features| features.as_array())
-        .map(|features| features.contains(&toml::Value::String("abi".to_string())))
-        .unwrap_or(false)
-    {
-        anyhow::bail!("Unable to generate ABI: NEAR SDK \"abi\" feature is not enabled")
+        .and_then(|deps| deps.get("near-sdk"))
```

Comment:
This won't catch
 - `near-sdk = { version = "..", features = ["abi"] }`
 - `near = { package = "near-sdk", version = ".." }`

and outright fail for `near-sdk = { git = ".." }`

Reply:
> near-sdk = { version = "..", features = ["abi"] }

What do you mean? This is absolutely being caught. We have multiple existing tests covering this and similar scenarios.

> near = { package = "near-sdk", version = ".." }

This, most likely, is never (or at least not in the foreseeable future) going to be supported. The problem lies in `near-sdk` itself which expects the crate to be called `near_sdk`. See https://github.com/near/cargo-near/blob/main/integration-tests/tests/cargo/mod.rs#L137

> and outright fail for near-sdk = { git = ".." }

No, it won't. You can only specify a git dependency in a *table* which is handled in a different branch.

Reply:
> What do you mean? This is absolutely being caught. We have multiple existing tests covering this and similar scenarios.

It's actually not. From the current logic, when the SDK dep is a table, it skips version checking altogether. And only checks for the `abi` feature.

> No, it won't. You can only specify a git dependency in a table which is handled in a different branch.

Same situation.

Since the `abi` feature flag is no longer required to generate ABI, when this check gets removed, many false negatives would scale through this pretty easily.

And when the dep is a table with version, one could specify "4.0" and it won't report any error until it hits the first `cargo metadata` which reports a verbose error:

```toml
[dependencies]
near-sdk = { version = "4", features = ["abi"] }
```

```console
error: Error invoking `cargo metadata`. Your `Cargo.toml` file is likely malformed

Caused by:
    `cargo metadata` exited with an error: error: failed to select a version for `near-sdk`.
        ... required by package `adder v0.1.0 (/Users/miraclx/git/near/near-sdk-rs/examples/adder)`
    versions that meet the requirements `^4` (locked to 4.0.0) are: 4.0.0
    
    the package `adder` depends on `near-sdk`, with features: `abi` but `near-sdk` does not have these features.
    
    
    failed to select a version for `near-sdk` which could resolve this conflict
```

> This, most likely, is never (or at least not in the foreseeable future) going to be supported. The problem lies in near-sdk itself which expects the crate to be called near_sdk. See [https://github.com/near/cargo-near/blob/main/integration-tests/tests/cargo/mod.rs#L137](https://github.com/near/cargo-near/blob/main/integration-tests/tests/cargo/mod.rs?rgh-link-date=2022-08-31T08%3A57%3A14Z#L137)

fair point.

Reply:
Ok, waiting for your PR then

Reply:
#49